### DISCUSSION TITLE
Seeed Studio XIAO SAMD21 with TinyUSB | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#346] [Q&A]
# Seeed Studio XIAO SAMD21 with TinyUSB | Seeed Studio Wiki

Seeed Studio XIAO SAMD21 with TinyUSB

https://wiki.seeedstudio.com/Seeeduino-XIAO-TinyUSB/

**Comment by jgrules2** [comment]
Hello,

Can you please update this sample? It seems that the version v010.5 is not working anymore for the Seeeduino XAIO SAMD21 when updating the core.

I have tried different setups even 1.6.1 from the board versions but every combination is giving compile error. One is missing the _Core.h and another tells 'TUD_MIDI_DESC_HEAD_LEN' was not declared in this scope.


Best Regards

**Reply by MatthewJeffson** [reply]
Hi, I have informed this issue to my colleague. I really appreciate you find it and your efforts to fix it.

Thanks!

**Comment by NNGB-kor** [comment]
Still this feature has not been fixed. / 2023/08/14
I can't use this little guy's features.

**Reply by MatthewJeffson** [reply]
We have came up with a new XIAO SAMD21 version as 1.8.4, to fix the issue. And here is the file: 
[1.8.4.zip](https://github.com/Seeed-Studio/wiki-documents/files/12340822/1.8.4.zip)
Our application engineer will test it according to the schedule. 

If possible, would you mind helping us to test it? by replacing the XIAO SAMD21 bootloader files as this new version on your computer.
Really appreciate!

**Reply by NNGB-kor** [reply]
Version 1.8.4 does not work with tinyUSB.
Version 1.7.9 is the last working version.
You guys don't think about code compatibility.
The fact that the tinyUSB feature doesn't work means there's very little use for this little board.
yes. We buy this for a reason.
That's because you can build the smallest HID device possible.
for example. Keyboard and joystick.
But... none of them work properly.
I miss the money I paid for this.
Sadly, we have to go back to the Leonardo board.
Or go for the more powerful teensy4.

**Comment by NNGB-kor** [comment]
To those who may be struggling with this shitty situation.

Use version 1.7.9 of Seed SAMD Baords.

The Adafruit TinyUSB library still uses 0.10.5.

The combination of the two should work. I used this to create my joystick environment.

**Reply by DragonAtlas** [reply]
Hi can you please share your code to see if I can get things working?

**Comment by MatthewJeffson** [comment]
I am terribly sorry for such inconvenience you guys have to experience. We will test out the right version of 1.8.4 and publish as soon as possible.
Regards!

**Reply by DragonAtlas** [reply]
Any updates?  I also have bought 3 boards and pretty useless without HID working

**Comment by DragonAtlas** [comment]
Hi All, I am having similar problem trying to make an HID scroll wheel using SAMD21 and an AS5600 magnetic encoder module.  Can someone help me with any example code that is working with HID?  NNGB-kor can you please share your joystick code?

**Comment by DragonAtlas** [comment]
Actually got my scroll to work by not including TinyUSB at all and just using 1.7.9 version of core as NNGB-kor suggested. If anyone interested here is my code.  Help from  Mauro Alfieri   
https://www.mauroalfieri.it/elettronica/as5600-knob-mouse-code.html

#include <Wire.h>
#include <Mouse.h>
#include <AS5600.h>
 

#ifdef ARDUINO_SAMD_VARIANT_COMPLIANCE
  #define SERIAL SerialUSB
  #define SYS_VOL   3.3
#else
  #define SERIAL Serial
  #define SYS_VOL   5
#endif

AMS_5600 ams5600;
int ang = 0;
long oldPosition  = -999;
void setup()
{
  SERIAL.begin(115200);
  Wire.begin();
  SERIAL.println(">>>>>>>>>>>>>>>>>>>>>>>>>>> ");
  if(ams5600.detectMagnet() == 0 ){
    while(1){
        if(ams5600.detectMagnet() == 1 ){
            SERIAL.print("Current Magnitude: ");
            SERIAL.println(ams5600.getMagnitude());
            break;
        }
        else{
            SERIAL.println("Can not detect magnet");
        }
        delay(1000);
    }
  }
}
/*******************************************************
/* Function: convertRawAngleToDegrees
/* In: angle data from AMS_5600::getRawAngle
/* Out: human readable degrees as float
/* Description: takes the raw angle and calculates
/* float value in degrees.
/*******************************************************/
float convertRawAngleToDegrees(word newAngle)
{
  /* Raw data reports 0 - 4095 segments, which is 0.087 of a degree */
  float retVal = newAngle * 0.02222222222;  //0.087890625 or 0.087912088 mycal for 360 degree 0.021978022/90degree. 0.010989010/45deg
  return (int) retVal;
}
void loop()
{
    long newPosition = convertRawAngleToDegrees(ams5600.getRawAngle());
    if (newPosition != oldPosition) {
      if ( oldPosition > newPosition ) { Mouse.move(0,0,-1); }
      if ( oldPosition < newPosition ) { Mouse.move(0,0,1); }
      oldPosition = newPosition;
    }
    SERIAL.println(String(convertRawAngleToDegrees(ams5600.getRawAngle())));
}

**Comment by dmanla** [comment]
Was this ever fixed? Why is this tutorial still fundamentally broken?

**Reply by nfs0619** [reply]
Hi, thanks for your feedback! We’ve tested the tutorial, and it works as expected when following the steps closely. Could you please try running through the tutorial again step by step? I’ve attached two screenshots showing that I successfully completed the tutorial, which might help clarify the process.

<img width="600" alt="企业微信截图_17401042124538" src="https://github.com/user-attachments/assets/30d3ffac-54c3-449e-90aa-46795174e343" />
<img width="600" alt="企业微信截图_17401042308016" src="https://github.com/user-attachments/assets/48cdf79c-1ed5-426d-ba9b-99775d39bc52" />

If you're still facing any issues, please don’t hesitate to reach out to our technical support team at [techsupport@seeed.io](mailto:techsupport@seeed.io). We’d be happy to assist you further.

**Reply by dmanla** [reply]
I am trying to use the mass-storage functionality of the TinyUSB library. None of the examples of mass-storage work.

The mass storage device appears, but clicking the drive causes File Explorer to freeze and sometimes crash.

**Reply by nfs0619** [reply]
Hi, thanks for your response!

The V1.0.0 and above versions of the TinyUSB library are not compatible with Seeed Studio XIAO SAMD21, and the mass-storage functionality you mentioned is likely a feature introduced in newer versions of the TinyUSB library, which may not work as expected. We recommend using version V0.10.5 of the TinyUSB library, as it is compatible with XIAO SAMD21 and should work properly.

If you have any other questions or need further assistance, please feel free to reach out to our technical support at [techsupport@seeed.io](mailto:techsupport@seeed.io)!

**Reply by dmanla** [reply]
Mass Storage has existed as a functionality in TinyUSB since before 0.10.5. I have tried the examples using 0.10.5. as well as the version and examples included with the samd21 board package provided by Seeed. Both have the same behavior I previously described when using the XAIO SAMD21 board.

I have contacted your support team and am awaiting a response.
